### DISCUSSION TITLE
Why does Mito require (mito:deftable ... ) in order to query the database?

### DISCUSSION THREAD

**Discussion by aykaramba** [fukamachi/mito#163] [General]
I am going through this tutorial to learn how to use the basics of Mito: https://dev.to/rajasegar/mito-an-orm-for-common-lisp-1n8n

1) The tutorial expects a user to do the following:

a) Connect to db with (mito:connect-toplevel ... )

b) Define table with (mito:deftable ...)

c) Create tables with (mito:ensure-table-exists ...)

... and so on. 

2) All of the above makes sense when setting up an app for the first time.  However, in order for me to be able to query the db with (mito:select-dao 'tablename) I cannot only do this: (mito:connect-toplevel ... )

Instead, in the app, I have to do the following:
(mito:connect-toplevel ... )
(mito:deftable ...)

Why is that?  Why cannot I just do (mito:connect-toplevel ...) and then automatically start doing queries on the db with (mit:select-dao 'tablename ...)?  What is the canonincal way to start performing queries on a connected db with mito because having to run the following code that defines a table every time seems wrong:

```
(mito:deftable warehouses ()
   ((location :col-type (:varchar 50))
   (capacity :col-type (:integer)))))
```

My understanding is that instantiates a CLOS object and allows us to (mito:ensure-table-exists ...).  Is there an equivalent canonical way to tell Mito that we are going to do a query on a table that is different from that? 

Appreciate any insights, really enjoying using Mito, it simplifies a whole lot of things.

**Comment by vindarel** [comment]
If we write straight on

    (mit:select-dao 'tablename ...)

we'll get that the `'tablename` symbol is unbound. As you said, `deftable` wil create a class/type, its symbols etc. 

Did you look at SxQL? https://github.com/fukamachi/sxql The SQL generator used by Mito, with which we compose queries. However it isn't as elegant as `(mit:select-dao 'tablename ...)`. 

my 2c really, as I didn't try this usecase. When working on an existing DB I went with either SxQL either with pure SQL in strings (using cl-dbi for the connection and passing parameters).

**Reply by aykaramba** [reply]
ohhhh ... is mito really meant for setting up databases and migrations and not usage?

It is really really convenient, the abstractions are incredibly sensible imho.

By the way, thank you for all of your Lisp coursework, it has really helped me!

**Reply by vindarel** [reply]
I would say yes for usage: one has to define the table with `deftable` and this gives us `find-dao`, `select-dao` and friends. But when one wants to compose queries, we use the SxQL syntax (once we want to write `when` clauses for example). 

When we work with an exsting DB, I don't know what's the best workflow, but I think you found it: define the table. Mito isn't written (yet, and AFAIK) to guess an existing DB's internals.

thanks for your feedback, appreciated. (and for taking the course!)

**Reply by vindarel** [reply]
Did I understand correctly and you'd like to use Mito with an existing DB you didn't create with Mito at the first place?

**Reply by aykaramba** [reply]
I want to be careful with how I answer this question because I don't want my current completely uneducated opinion to influence the direction of Mito, so let me answer it this way:

I thought Mito and SxQL were two different Lispy abstractions over SQL.  I thought Mito was an abstraction that was closer to using CLOS objects and slots and that I would have to use Lisp constructs to perform queries while SxQL is making SQL Lispy.

As I have taken your Lisp course and made a couple of tiny apps with a db back end, I found composing pure sql a bit messy.  SxQL provides a way to write SQL style queries using symbols and Mito was taking an object oriented approach.

I don't know how much of that is correct and that is just the impression I got as a noob looking to get data in and out of SQLite.  Mito just seemed so super convenient, but I understand that SQL like queries are based on an area of mathematical study and reinventing the wheel has it's tradeoffs.

The above would be the reason why I tried to use Mito instead of SxQL and just like with SxQL my impression was that I would be able to use it with any db that was already created.

// edit

And I now understand that Mito is used in conjunction with SxQL.  I clearly thought they were supposed to do the same job.  I guess we can do things like this with our code (via this tutorial: https://dev.to/rajasegar/mito-an-orm-for-common-lisp-1n8n):

```
(loop for x in (mito:select-dao 'warehouses (sxql:order-by :id))
		        collect (slot-value x 'location) )
```

I will work through your Mito/SxQL tutorial later this week.
